File: Data Structures/Array/Multi-dimensional/Python3/Largest Square.py
"""
             ...........................................................
               Solver       :   Arif Khan
               Problem      :   Largest Square
               Level        :   Medium
               Judge        :   HackerEarth
               Date         :   29/11/2023
               Time         :   O(N)
               Memory       :   3.5KB
               Difficulty   :   Easy
               problem_link :   https://www.hackerearth.com/practice/data-structures/arrays/multi-dimensional/practice-problems/algorithm/largest-square-3d7a938a/
               github       :   https://github.com/arifkhan1990
               portfolio    :   https://arif-portfolio-orcin.vercel.app/
               gamil        :   arifkhanshubro@gmail.com
               phone        :   01960518551
             ...........................................................
"""
def squared_distance(point1, point2):
    return (point1[0] - point2[0])**2 + (point1[1] - point2[1])**2

def find_largest_square(points, n):
    point_set = set(points)

    max_side_length = 0
    result_points = None

    for i in range(n - 1):
        for j in range(i + 1, n):
            if points[i][0] != points[j][0] and points[i][1] != points[j][1]:
                diagonal1 = (points[j][0], points[i][1])
                diagonal2 = (points[i][0], points[j][1])

                if diagonal1 in point_set and diagonal2 in point_set:
                    side_length = squared_distance(points[i], points[j])
                    if side_length > max_side_length:
                        max_side_length = side_length
                        result_points = (points[i], points[j], diagonal1, diagonal2)

    return result_points

# Take user input for points
n = int(input())
points = []

for _ in range(n):
    x, y = map(int, input().split())
    points.append((x, y))

result = find_largest_square(points, n)

if result:
    bottom_left_point = min(result, key=lambda x: (x[1], x[0]))
    print(bottom_left_point[0], bottom_left_point[1])
else:
    print("-1")

File: Data Structures/Array/Multi-dimensional/C++/Largest Square.cpp

/*
             ...........................................................
               Solver       :   Arif Khan
               Problem      :   Largest Square
               Level        :   Medium
               Judge        :   HackerEarth
               Date         :   29/11/2023
               Time         :   O(N)
               Memory       :   3.5KB
               Difficulty   :   Easy
               problem_link :   https://www.hackerearth.com/practice/data-structures/arrays/multi-dimensional/practice-problems/algorithm/largest-square-3d7a938a/
               github       :   https://github.com/arifkhan1990
               portfolio    :   https://arif-portfolio-orcin.vercel.app/
               gamil        :   arifkhanshubro@gmail.com
               phone        :   01960518551
             ...........................................................
*/

#include <bits/stdc++.h>
using namespace std;

double squared_distance(pair<int, int> point1, pair<int, int> point2) {
    return pow(point1.first - point2.first, 2) + pow(point1.second - point2.second, 2);
}

pair<int, int> find_bottom_left_point(const vector<pair<int, int>>& square_points) {
    return *min_element(square_points.begin(), square_points.end(), [](const auto& p1, const auto& p2) {
        return (p1.second == p2.second) ? (p1.first < p2.first) : (p1.second < p2.second);
    });
}

vector<pair<int, int>> find_largest_square(const vector<pair<int, int>>& points) {
    int n = points.size();
    set<pair<int, int>> point_set(points.begin(), points.end());

    double max_side_length = 0;
    vector<pair<int, int>> result_points;

    for (int i = 0; i < n - 1; ++i) {
        for (int j = i + 1; j < n; ++j) {
            if (points[i].first != points[j].first && points[i].second != points[j].second) {
                pair<int, int> diagonal1 = {points[j].first, points[i].second};
                pair<int, int> diagonal2 = {points[i].first, points[j].second};

                if (point_set.count(diagonal1) && point_set.count(diagonal2)) {
                    double side_length = squared_distance(points[i], points[j]);
                    if (side_length > max_side_length) {
                        max_side_length = side_length;
                        result_points = {points[i], points[j], diagonal1, diagonal2};
                    }
                }
            }
        }
    }

    return result_points;
}

int main() {
    int n;
    cin >> n;

    vector<pair<int, int>> points;
    for (int i = 0; i < n; ++i) {
        int x, y;
        cin >> x >> y;
        points.emplace_back(x, y);
    }

    vector<pair<int, int>> result = find_largest_square(points);

    if (!result.empty()) {
        pair<int, int> bottom_left_point = find_bottom_left_point(result);
        cout << bottom_left_point.first << " " << bottom_left_point.second << endl;
    } else {
        cout << "-1" << endl;
    }

    return 0;
}
